### PR TITLE
12.0 t10293 cancel reservations

### PR DESCRIPTION
**Pull request by robinkeunen**
## Checklist before approval

- [x] Tests are present (or not needed).
- [x] Credits/copyright have been changed correctly.
- [ ] Change log snippet is present.
- [x] (If a new module) Moving this to OCA has been considered.

### CODE REVIEW

File: resource_activity/models/resource_activity.py [hunk 1]
```diff
@@ -341,7 +341,7 @@ def _compute_duration(self):
 
     @api.multi
     def search_all_resources(self):
-        self.mapped("registrations").search_resources()
+        return self.mapped("registrations").search_resources()
```
Comment (huguesdk) [inline resource_activity/models/resource_activity.py:350]: why is `return` needed?

File: resource_activity/models/resource_activity_registration.py [hunk 2]
```diff
@@ -199,15 +199,19 @@ def search_resources(self):
...
                     else:
                         registration.state = "waiting"
-                        # fixme create own cursor to persist waiting state
-                        self.env.cr.commit()  # pylint: disable=invalid-commit
-                        raise UserError(
-                            _("Not enough resource found for the registration")
-                        )
+                        return {
+                            "type": "ir.actions.act_window.message",
+                            "title": _("Réservation Impossible"),
```
Comment (huguesdk) [inline resource_activity/models/resource_activity_registration.py]:
```suggestion
                            "title": _("Registration Not Possible"),
```

File: resource_activity/models/resource_available.py [hunk 3]
```diff
@@ -34,40 +39,38 @@ class ResourceAvailable(models.Model):
 
     @api.multi
     def action_reserve(self):
-        for resource_available in self.filtered(lambda record: record.state == "free"):
```
Comment (huguesdk) [inline resource_activity/models/resource_available.py:37]: how come there is no need to filter on the `free` state anymore?

File: resource_activity/models/resource_available.py [hunk 4]
```diff
@@ -34,40 +39,38 @@ class ResourceAvailable(models.Model):
...
+                    "date_start": available_resource.registration_id.date_start,
+                    "date_end": available_resource.registration_id.date_end,
+                    "date_lock": available_resource.registration_id.date_lock,
+                    "state": available_resource.registration_id.booking_type,
+                    "location": available_resource.registration_id.location_id.id,
+                    "activity_registration_id": available_resource.registration_id.id,
+                }
+            )
+            available_resource.state = "selected"
+            available_resource.activity_id.registrations.action_refresh()
```
Comment (huguesdk) [inline resource_activity/models/resource_available.py]: this should not be needed with correct dependencies on computed fields. not blocking as it was already there, and we’re not fixing all issues here.

File: resource_activity/tests/test_resource_activity.py [hunk 5]
```diff
@@ -227,3 +227,145 @@ def test_no_update_resource_registration_date_end_when_action_done_after_date_en
...
+        mtb_1_allocation = self.env["resource.allocation"].create(
+            {
+                "resource_id": self.mtb_1.id,
+                "date_start": date_start,
+                "date_end": date_end,
+                # "location_id": self.main_location.id,
+                # "partner_id": self.partner_demo.id,
+            }
+        )
+        mtb_1_allocation.action_confirm()
```
Comment (huguesdk) [inline resource_activity/tests/test_resource_activity.py]: why is this needed? this means that on line 305, `mtb_1` will be linked to 2 different allocations. why not simply change the `resource_id` of the allocation to another available resource?

File: resource_activity/models/resource_activity.py [hunk 6]
```diff
@@ -350,13 +350,11 @@ def reserve_needed_resource(self):
...
-        registrations = self.env["resource.activity.registration"].browse()
-        for activity in self:
-            for registration in activity.registrations:
-                if registration.state == "booked":
-                    registrations |= registration
-        registrations.action_cancel()
-        registrations.action_draft()
+        booked_registrations = self.mapped("registrations").filtered(
+            lambda r: r.state == "booked"
+        )
```
Comment (huguesdk) [inline resource_activity/models/resource_activity.py:361]: much more understandable. 👍

File: resource_activity/models/resource_activity_registration.py [hunk 7]
```diff
@@ -199,15 +199,19 @@ def search_resources(self):
...
+                            "message": _(
+                                "Not enough resource found for the registration"
+                            ),
+                        }
```
Comment (huguesdk) [inline resource_activity/models/resource_activity_registration.py]: oh yeah! 🎉

File: resource_activity/models/resource_activity_registration.py [hunk 8]
```diff
@@ -228,31 +232,33 @@ def action_refresh(self):
...
-                    if (
-                        registration.quantity_needed - registration.quantity_allocated
-                        <= 0
-                    ):
-                        break
-                    resource_available.action_reserve()
-
-                (registration.resource_activity_id.registrations.action_refresh())
-        return True
+                return
```
Comment (huguesdk) [inline resource_activity/models/resource_activity_registration.py]:
```suggestion
                continue
```

File: resource_activity/models/resource_activity_registration.py [hunk 9]
```diff
@@ -228,31 +232,33 @@ def action_refresh(self):
...
+            free_resources = registration.resources_available.filtered(
+                lambda record: record.state == "free"
+            )
+            free_resources[: registration.quantity_needed].action_reserve()
+
+            if registration.quantity_allocated >= registration.quantity_needed:
+                registration.state = registration.booking_type
+
+            registration.resource_activity_id.registrations.action_refresh()
+            return
```
Comment (huguesdk) [inline resource_activity/models/resource_activity_registration.py]: the `return` statement is in the `for` loop, which means it will never reach the 2nd element of the loop.

File: resource_activity/models/resource_activity_registration.py [hunk 10]
```diff
@@ -228,31 +232,33 @@ def action_refresh(self):
...
+
+            # update available resources
+            registration.search_resources()
+            free_resources = registration.resources_available.filtered(
+                lambda record: record.state == "free"
+            )
+            free_resources[: registration.quantity_needed].action_reserve()
+
+            if registration.quantity_allocated >= registration.quantity_needed:
+                registration.state = registration.booking_type
```
Comment (huguesdk) [inline resource_activity/models/resource_activity_registration.py:252]: a comment would be nice, especially for the 2nd line, but also to explain how it is possible for `quantity_allocated` to be `>` than `quantity_needed`.

File: resource_activity/models/resource_activity_registration.py [hunk 11]
```diff
@@ -228,31 +232,33 @@ def action_refresh(self):
...
-                    ):
-                        break
-                    resource_available.action_reserve()
-
-                (registration.resource_activity_id.registrations.action_refresh())
-        return True
+                return
+
+            # update available resources
+            registration.search_resources()
```
Comment (huguesdk) [inline resource_activity/models/resource_activity_registration.py:242]: why is an extra search needed here compared to before?

File: resource_activity/models/resource_activity_registration.py [hunk 12]
```diff
@@ -228,31 +232,33 @@ def action_refresh(self):
...
+            registration.resource_activity_id.registrations.action_refresh()
+            return
 
     @api.multi
     def action_cancel(self):
         for registration in self:
-            for resource_available in registration.resources_available:
-                resource_available.action_cancel()
+            registration.allocations.action_cancel()
+            registration.resources_available.unlink()
```
Comment (huguesdk) [inline resource_activity/models/resource_activity_registration.py]: deleting the `resources_available` is a functional change. why not keep the existing mechanism, now that the correct links are in place?

File: resource_activity/models/resource_activity_registration.py [hunk 13]
```diff
@@ -159,6 +92,73 @@ def _compute_state(self):
...
+
+            # if new registrations goes to booked or option state
+            elif (
+                registration.resource_activity_id.sale_orders
+                and registration.state in ["booked", "option"]
+            ):
+                registration.need_push = True
+
+    @api.multi
+    @api.depends("allocations.state")
```
Comment (huguesdk) [inline resource_activity/models/resource_activity_registration.py:141]:
```suggestion
    @api.depends("allocations", "allocations.state")
```
like this, it would also work if an allocation is deleted.

File: resource_activity/models/resource_activity_registration.py [hunk 14]
```diff
@@ -159,6 +92,73 @@ def _compute_state(self):
...
+    def _compute_quantity_allocated(self):
+        for registration in self:
+            registration.quantity_allocated = len(
+                registration.allocations.filtered(lambda a: a.state != "cancel")
+            )
+
+    @api.multi
+    @api.depends("allocations")
```
Comment (huguesdk) [inline resource_activity/models/resource_activity_registration.py]:
```suggestion
    @api.depends("allocations", "allocations.state")
```
it actually depends on `quantity_allocated`, but (if i’m not mistaken), the dependencies are not recursively computed.

File: resource_activity/models/resource_available.py [hunk 15]
```diff
@@ -12,9 +12,14 @@ class ResourceAvailable(models.Model):
...
+    resource_id = fields.Many2one(
+        comodel_name="resource.resource", string="Resource", required=True
+    )
     registration_id = fields.Many2one(
-        "resource.activity.registration", string="Registration"
+        comodel_name="resource.activity.registration", string="Registration"
+    )
+    allocation_id = fields.Many2one(
+        comodel_name="resource.allocation", string="Allocation"
     )
```
Comment (huguesdk) [inline resource_activity/models/resource_available.py:52]: to manage existing records, a migration script should set this field from the `resource_id` and `registration_id`.

File: resource_activity/models/resource_available.py [hunk 16]
```diff
@@ -12,9 +12,14 @@ class ResourceAvailable(models.Model):
     _description = "Resource Available"
 
     name = fields.Char(related="resource_id.serial_number", string="Name")
-    resource_id = fields.Many2one("resource.resource", string="Resource", required=True)
+    resource_id = fields.Many2one(
+        comodel_name="resource.resource", string="Resource", required=True
+    )
```
Comment (huguesdk) [inline resource_activity/models/resource_available.py]: this should be a stored (settable) computed field that depends on `allocation_id` and `allocation_id.resource_id`. it would allow to automatically update the `resource_id` when it is change in the allocation.

same remark for the `state` field.

File: resource_planning/wizard/allocate_resource_wizard.py [hunk 17]
```diff
@@ -82,13 +82,22 @@ def search_resources(self):
...
-                self.date_start,
-                self.date_end,
-                self.partner_id,
-                self.location,
-                self.date_lock,
+    def book_resources(self):
+        self.ensure_one()
+        for resource in self.resources:
+            # cf resource.check_availabilities(
+            #  this function returns resource.available records
```
Comment (huguesdk) [inline resource_planning/wizard/allocate_resource_wizard.py:89]: `resource.check_availabilities()` returns `resource.resource` records. i’m not sure i understand the comment.

File: resource_planning/wizard/allocate_resource_wizard.py [hunk 18]
```diff
@@ -82,13 +82,22 @@ def search_resources(self):
...
+            #  nothing was done with it here
+            #  hope book_resources is called right after search_resources
```
Comment (huguesdk) [inline resource_planning/wizard/allocate_resource_wizard.py:91]: doesn’t the `resource.allocation._check_resource_allocations()` ensures that the resource is still available?

**Review by huguesdk — CHANGES_REQUESTED** [review]
wow, quite some changes! great tests! thanks for all this.

File: resource_activity/models/resource_activity.py (hunk 1 above)
Comment (robinkeunen) [reply resource_activity/models/resource_activity.py:350]: Je ne vois pas de raison dans le code actuel ni au moment du commit pour le return. A part que c'est si stateful que je _cringe_.

File: resource_activity/models/resource_available.py (hunk 3 above)
Comment (robinkeunen) [reply resource_activity/models/resource_available.py:37]: Mmmh, I think it might be safer to leave it there. I might have overlooked that in self, there might be resource_available in other states (as the model name does not suggest).

File: resource_activity/models/resource_available.py (hunk 4 above)
Comment (robinkeunen) [reply resource_activity/models/resource_available.py]: I agree.

File: resource_activity/tests/test_resource_activity.py (hunk 5 above)
Comment (robinkeunen) [reply resource_activity/tests/test_resource_activity.py]: I think you're right, booking `mtb_1` is not needed for the test.

File: resource_activity/models/resource_activity_registration.py (hunk 14 above)
Comment (robinkeunen) [reply resource_activity/models/resource_activity_registration.py]: Then `allocations.state` is enough. It indeed also depends on `quantity_allocated` and `quantity_needed`(see comment at L235) but I stopped refactoring after overshooting the estimated time (plus this code will be dropped at migration).

File: resource_planning/wizard/allocate_resource_wizard.py (hunk 17 above)
Comment (robinkeunen) [reply resource_planning/wizard/allocate_resource_wizard.py:89]:
```py
# cf resource.check_availabilities(
#  this function returns resource.available records
#  nothing was done with it here
```

I think that comment belongs to the commit message. I'm explaining why I create the allocation directly rather than using check_availabilities.

```py
# hope book_resources is called right after search_resources
```

This function is called from a button shown after the button search_resources was clicked. We therefore hope the allocated resource is still free. It's a bit risky but I stopped refactoring.

**Comment by huguesdk** [comment]
@robinkeunen would like to review my changes?

File: resource_activity/migrations/12.0.1.4.0/post-set_resource_available_allocation.py [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+            ral.id,
+            'selected'
+        from resource_allocation as ral
+        left join resource_available as rav on
+            ral.id = rav.allocation_id
+        where
+            ral.activity_registration_id is not null and
+            ral.state <> 'cancel' and
+            rav.allocation_id is null
+        order by 3
```
Comment (robinkeunen) [inline resource_activity/migrations/12.0.1.4.0/post-set_resource_available_allocation.py]: 👍

File: resource_activity/migrations/12.0.1.4.0/post-set_resource_available_allocation.py [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+    # now, update the quantity_allocated field of registrations.
+    cr.execute(
+        """
+        with resource_allocation_count as (
+            select rar.id, count(0) as quantity
+            from resource_activity_registration as rar
+            inner join resource_allocation as ral on
+                ral.activity_registration_id = rar.id
+            where
+                ral.state <> 'cancel'
```
Comment (robinkeunen) [inline resource_activity/migrations/12.0.1.4.0/post-set_resource_available_allocation.py]: I think you should let the allocation in state `option` out of the count. In state "maintenance" as well to be safe but I'm not sure how it relates to the registrations.

Actually : `ral.state = 'booked'`.

File: resource_activity/models/resource_activity_registration.py [hunk 21]
```diff
@@ -138,15 +138,25 @@ def _compute_need_push(self):
                 registration.need_push = True
 
     @api.multi
-    @api.depends("allocations.state")
+    @api.depends(
+        "allocations",
```
Comment (robinkeunen) [inline resource_activity/models/resource_activity_registration.py]: I'm pretty sure this is not needed.

File: resource_activity/models/resource_activity_registration.py [hunk 22]
```diff
@@ -138,15 +138,25 @@ def _compute_need_push(self):
...
+        "allocations.state",
+        "resources_available.allocation_id",
```
Comment (robinkeunen) [inline resource_activity/models/resource_activity_registration.py]: Same here, if the allocation changes, Odoo considers the state changes. Might be worth checking though.

File: resource_activity/models/resource_activity_registration.py [hunk 23]
```diff
@@ -138,15 +138,25 @@ def _compute_need_push(self):
...
             )
 
     @api.multi
-    @api.depends("allocations")
+    @api.depends(
+        "allocations",
+        "allocations.state",
+        "resources_available.allocation_id",
+        "resources_available.allocation_id.state",
+    )
```
Comment (robinkeunen) [inline resource_activity/models/resource_activity_registration.py]: idem.

File: resource_activity/models/resource_activity_registration.py [hunk 24]
```diff
@@ -201,7 +211,7 @@ def search_resources(self):
                         registration.state = "waiting"
                         return {
                             "type": "ir.actions.act_window.message",
-                            "title": _("Réservation Impossible"),
+                            "title": _("Registration Not Possible"),
```
Comment (robinkeunen) [inline resource_activity/models/resource_activity_registration.py:204]: Will need translation.

File: resource_activity/models/resource_available.py [hunk 25]
```diff
@@ -11,9 +12,37 @@ class ResourceAvailable(models.Model):
     _name = "resource.available"
     _description = "Resource Available"
 
+    @api.depends("allocation_id", "allocation_id.resource_id")
+    def _compute_resource_id(self):
+        for record in self:
```
Comment (robinkeunen) [inline resource_activity/models/resource_available.py:17]: ah, t'es finalement convaincu par les `for record in self` toi ?

**Review by robinkeunen — COMMENTED** [review]
A few comment, can't approve my own PR though.

File: resource_activity/migrations/12.0.1.4.0/post-set_resource_available_allocation.py (hunk 20 above)
Comment (huguesdk) [reply resource_activity/migrations/12.0.1.4.0/post-set_resource_available_allocation.py]: i also had doubts about this, but what i understood from the code is that `option` and `maintenance` states are also considered as a valid resource allocation (meaning that it is not available for others). `maintenance` is not used in the normal registration flow, and there are no `resource.allocation` records with this state that are linked to a registration. i’ve just checked and there are no `resource.allocation` records with the `option` state.

File: resource_activity/models/resource_available.py (hunk 25 above)
Comment (huguesdk) [reply resource_activity/models/resource_available.py:17]: do you have any context on this? the odoo examples use it.

File: resource_activity/models/resource_activity_registration.py (hunk 22 above)
Comment (huguesdk) [reply resource_activity/models/resource_activity_registration.py]: indeed, you’re right. i made some tests and only the deepest field is needed.

**Review by huguesdk — APPROVED** [review]
thanks @robinkeunen for the final review. i’ve removed the unnecessary dependencies and squashed the commits.

File: resource_activity/models/resource_available.py (hunk 25 above)
Comment (robinkeunen) [reply resource_activity/models/resource_available.py:17]: I think they (Odoo) like to have the same loop everywhere, I guess it allows to have autogenerated loop code easier. Personally, I prefer to know what I'm looping on and have a clearer name.

Comment (huguesdk) [reply resource_activity/models/resource_available.py:17]: it’s a matter of taste, i guess. i prefer a more generic name, like with `self`. with a different name, i’m always asking myself whether we’re referring to something else.
